pkg/secrets-store/utils.go

```diff
@@ -0,0 +1,790 @@
+
+	deleteStatusFn := func() (bool, error) {
+		item, podNS, err := getItemWithPodID(ctx, podUID)
+		if err == nil && len(podNS) > 0 {
+			secretProviderClass = item.GetName()
+			namespace = podNS
+			if err = deleteStatus(ctx, item, podUID); err != nil {
+				return false, nil
+			}
+			return true, nil
+		}
+		return true, nil
+	}
+
+	if err := wait.ExponentialBackoff(wait.Backoff{
+		Steps:    5,
+		Duration: 1 * time.Millisecond,
+		Factor:   1.0,
+		Jitter:   0.1,
+	}, deleteStatusFn); err != nil {
+		log.Error(err, "max retries for deleting status reached for pod: %s, ns: %s", podUID, namespace)
+		return err
+	}
+
+	if len(namespace) > 0 && len(secretProviderClass) > 0 {
+		deleteSecretFn := func() (bool, error) {
+			item, err := getSecretProviderItemByName(ctx, secretProviderClass)
+			if err != nil {
+				log.Errorf("failed to get secret provider item, err: %v for pod: %s, ns: %s", err, podUID, namespace)
+				return false, nil
+			}
+			count := getStatusCount(item)
+			// only delete when no more pods are associated with it
+			if count == 0 {
+				///TODO: we assume all files are mounted from a single secretsproviderclass
+				/// a pod could have multiple volumes pointing to diff secretproviderclass objs
+				for _, s := range secretObjects {
+					secretObject, ok := s.(map[string]interface{})
+					if !ok {
+						continue
+					}
+
+					secretName, err := getStringFromObject(secretObject, "secretName")
+					if err != nil {
+						continue
+					}
+					err = deleteK8sSecret(ctx, secretName, namespace)
+					if err != nil {
+						return false, nil
+					}
+				}
+			}
+			return true, nil
+		}
+
+		if err := wait.ExponentialBackoff(wait.Backoff{
+			Steps:    5,
+			Duration: 1 * time.Millisecond,
+			Factor:   1.0,
+			Jitter:   0.1,
+		}, deleteSecretFn); err != nil {
+			log.Error(err, "max retries for deleting secret reached for pod: %s, ns: %s", podUID, namespace)
+			return err
+		}
+	}
+	return nil
+}
+
+// createOrUpdateK8sSecret creates or updates a K8s secret with data from mounted files
+// If a secret with the same name already exists in the namespace of the pod, it's updated.
+func createOrUpdateK8sSecret(ctx context.Context, name string, namespace string, datamap map[string][]byte, secretType corev1.SecretType) error {
+	// recreating client here to prevent reading from cache
+	c, err := getClient()
+	if err != nil {
+		return err
+	}
+	secretKey := types.NamespacedName{
+		Namespace: namespace,
+		Name:      name,
+	}
+
+	secret := &corev1.Secret{
+		ObjectMeta: metav1.ObjectMeta{
+			Namespace: namespace,
+			Name:      name,
+		},
+		Type: secretType,
+	}
+
+	err = c.Get(ctx, secretKey, secret)
+	if err != nil {
+		log.Error(err, "error from c.Get for secret: %s, ns: %s", name, namespace)
+		if errors.IsNotFound(err) {
+			secret.Data = datamap
+			if err := c.Create(ctx, secret); err != nil {
+				log.Error(err, "error while creating K8s secret: %s, ns: %s", name, namespace)
+				return err
+			}
```

Comment:
there should be a `return nil` here if error is nil, otherwise this will also go through the update

Reply:
Good catch!